modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java

```diff
@@ -3216,196 +3207,227 @@ private void body0() throws InterruptedException, IgniteCheckedException {
+                    if (isCancelled())
+                        break;
 
-                        if (task instanceof RebalanceReassignExchangeTask)
-                            exchId = ((RebalanceReassignExchangeTask) task).exchangeId();
-                        else if (task instanceof ForceRebalanceExchangeTask) {
-                            forcePreload = true;
+                    if (task instanceof RebalanceReassignExchangeTask) {
+                        RebalanceReassignExchangeTask reassignTask = (RebalanceReassignExchangeTask)task;
+
+                        exchId = reassignTask.exchangeId();
+
+                        GridDhtPartitionsExchangeFuture lastFut = lastFinishedFut.get();
+
+                        if (lastFut != null) {
+                            AffinityTopologyVersion lastAffChangedVer =
+                                cctx.exchange().lastAffinityChangedTopologyVersion(lastFut.topologyVersion());
+
+                            // Reassign request as part of the last finished exchange.
+                            if (exchId.topologyVersion().isBetween(lastFut.initialVersion(), lastFut.topologyVersion())) {
+                                if (lastFut.hasInapplicableNodesForRebalance())
+                                    exchFut = lastFut;
+                            }
+                            else  if (lastAffChangedVer.after(exchId.topologyVersion())) {
+                                    // There is a new exchange which should trigger rebalancing.
+                                    // This reassign request can be skipped.
+                                if (log.isInfoEnabled()) {
+                                    log.info("Partitions reassign request skipped due to affinity was already changed" +
+                                        " [reassignTopVer=" + exchId.topologyVersion() +
+                                        ", lastAffChangedTopVer=" + lastAffChangedVer +']');
+                                }
 
-                            timeout = 0; // Force refresh.
+                                continue;
+                            }
+                            else {
+                                // There was an exchange that does not change the affinity.
+                                for (GridDhtPartitionsExchangeFuture f : cctx.exchange().exchangeFutures()) {
+                                    if (f.isDone() && f.topologyVersion().equals(lastAffChangedVer)) {
+                                        exchFut = f.hasInapplicableNodesForRebalance()? f : null;
 
-                            exchId = ((ForceRebalanceExchangeTask)task).exchangeId();
+                                        break;
+                                    }
+                                }
+                            }
                         }
-                        else {
-                            assert task instanceof GridDhtPartitionsExchangeFuture : task;
+                    }
```

Comment:
This is a lot of code.
I think it can be simplified (if you will pass exchange future in reassigntask)

 to:

```
if (task instanceof RebalanceReassignExchangeTask) {
                            exchId = ((RebalanceReassignExchangeTask) task).exchangeId();

                            GridDhtPartitionsExchangeFuture fut = ((RebalanceReassignExchangeTask) task).future();
                            assert fut.changedAffinity();                            

                            if (fut.hasInapplicableNodesForRebalance()) {
                                  GridDhtPartitionsExchangeFuture lastFut = lastFinishedFut.get();

                                  AffinityTopologyVersion lastAffVer = cctx.exchange()
                                      lastAffinityChangedTopologyVersion(lastFut.topologyVersion());
                            
                                 if (fut.exchangeId().topologyVersion().equals(lastAffVer))
                                      exchFut = fut;
                             }
}
```

I also suggest to avoid iterations over _cctx.exchange().exchangeFutures()_. 
I have plans to remove this method in the future (because we do not need exchange history)

Reply:
Improved in the way as you suggested.